<a href="https://colab.research.google.com/github/Nat2364/Assignment1/blob/main/Assignment1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
import zipfile
import os

zip_path = '/content/drive/My Drive/UniversityAssignmentCrimeData.zip'
output_dir = './crime_data'

os.makedirs(output_dir, exist_ok=True)

with zipfile.ZipFile(zip_path, 'r') as zip_ref:
    zip_ref.extractall(output_dir)

print(f"'{zip_path}' extracted to '{output_dir}'")

'/content/drive/My Drive/UniversityAssignmentCrimeData.zip' extracted to './crime_data'


In [4]:
FILENAME_SUFFIX = '-west-midlands-street.csv'

In [5]:
all_files = []

for root, dirs, files in os.walk(output_dir):
    for file in files:
        if file.endswith(FILENAME_SUFFIX):
            full_path = os.path.join(root, file)
            all_files.append(full_path)

print("Identified CSV files:")
for f in all_files:
    print(f)


Identified CSV files:
./crime_data/2025-08/2025-08-west-midlands-street.csv
./crime_data/2024-11/2024-11-west-midlands-street.csv
./crime_data/2023-12/2023-12-west-midlands-street.csv
./crime_data/2024-05/2024-05-west-midlands-street.csv
./crime_data/2025-03/2025-03-west-midlands-street.csv
./crime_data/2025-05/2025-05-west-midlands-street.csv
./crime_data/2025-09/2025-09-west-midlands-street.csv
./crime_data/2024-12/2024-12-west-midlands-street.csv
./crime_data/2025-01/2025-01-west-midlands-street.csv
./crime_data/2024-01/2024-01-west-midlands-street.csv
./crime_data/2024-04/2024-04-west-midlands-street.csv
./crime_data/2024-06/2024-06-west-midlands-street.csv
./crime_data/2024-02/2024-02-west-midlands-street.csv
./crime_data/2025-02/2025-02-west-midlands-street.csv
./crime_data/2024-09/2024-09-west-midlands-street.csv
./crime_data/2025-04/2025-04-west-midlands-street.csv
./crime_data/2025-11/2025-11-west-midlands-street.csv
./crime_data/2025-06/2025-06-west-midlands-street.csv
./crim

## Load CSVs into DataFrames

### Subtask:
Iterate through the identified CSV files, reading each into a pandas DataFrame. Store these DataFrames in a list.


**Reasoning**:
To load the CSV files, I need to import the pandas library and then iterate through the list of identified file paths, reading each into a DataFrame and storing them in a list.



In [6]:
import pandas as pd

frames = []

for file_path in all_files:
    df = pd.read_csv(file_path)
    frames.append(df)

print(f"Loaded {len(frames)} DataFrames into the 'frames' list.")

Loaded 24 DataFrames into the 'frames' list.


## Concatenate DataFrames

### Subtask:
Combine all individual DataFrames from the list into one master DataFrame using `pd.concat`.


**Reasoning**:
To combine all individual DataFrames into one master DataFrame, I will use `pd.concat()` on the `frames` list and reset the index.



In [7]:
combined_df = pd.concat(frames, ignore_index=True)

print(f"Combined DataFrame shape: {combined_df.shape}")
print("First 5 rows of the combined DataFrame:")
print(combined_df.head())

Combined DataFrame shape: (668108, 12)
First 5 rows of the combined DataFrame:
                                            Crime ID    Month  \
0  1fc0f0d787b2f716930f49ff2eb9a49c55f327cf543aac...  2025-08   
1  626f61477f4c1c74fb96475907221d457ba73b887611bb...  2025-08   
2  1122d890c2bc0e57a44a8b6ba5518ea9e98ee25eaf76b7...  2025-08   
3  ca760e7e4a2d777c656807242187e8bf1b33e3457d222c...  2025-08   
4  0729cdd7e2bb71ce3388a94f1d18aff6ddcb244e4d9298...  2025-08   

            Reported by          Falls within  Longitude   Latitude  \
0  West Midlands Police  West Midlands Police  -1.847899  52.591564   
1  West Midlands Police  West Midlands Police  -1.847899  52.591564   
2  West Midlands Police  West Midlands Police  -1.847899  52.591564   
3  West Midlands Police  West Midlands Police  -1.847899  52.591564   
4  West Midlands Police  West Midlands Police  -1.849177  52.593202   

                       Location  LSOA code        LSOA name  \
0   On or near Royal Park Close  E010094

## Final Task

### Subtask:
Confirm that all 24 CSV files have been successfully loaded and merged into a single DataFrame, and confirm with the user that the data is ready for analysis.


## Summary:

### Q&A
Yes, all 24 CSV files have been successfully loaded and merged into a single DataFrame, and the data is confirmed to be ready for analysis.

### Data Analysis Key Findings
*   **File Identification:** A total of 24 CSV files matching the specified pattern (`-west-midlands-street.csv`) were successfully identified within the `./crime_data` directory.
*   **Data Loading:** All 24 identified CSV files were individually loaded into pandas DataFrames.
*   **Data Merging:** These 24 individual DataFrames were successfully concatenated into a single master DataFrame named `combined_df`.
*   **Combined DataFrame Dimensions:** The final `combined_df` contains `668,108` rows and `12` columns, indicating a large dataset ready for processing.
*   **Data Integrity Check:** The head of the `combined_df` was displayed, confirming that the data was correctly structured and preserved after the merge.

### Insights or Next Steps
*   The data is now prepared and ready for detailed exploratory data analysis, cleaning, or model building based on the crime data.
*   It would be beneficial to perform an initial data quality check, including examining data types, checking for missing values, and reviewing unique values in categorical columns.


In [8]:
combined_df

,Crime ID,Month,Reported by,Falls within,Longitude,Latitude,Location,LSOA code,LSOA name,Crime type,Last outcome category,Context
0,1fc0f0d787b2f716930f49ff2eb9a49c55f327cf543aac...,2025-08,West Midlands Police,West Midlands Police,-1.847899,52.591564,On or near Royal Park Close,E01009417,Birmingham 001A,Criminal damage and arson,Unable to prosecute suspect,NaN
1,626f61477f4c1c74fb96475907221d457ba73b887611bb...,2025-08,West Midlands Police,West Midlands Police,-1.847899,52.591564,On or near Royal Park Close,E01009417,Birmingham 001A,Drugs,Local resolution,NaN
2,1122d890c2bc0e57a44a8b6ba5518ea9e98ee25eaf76b7...,2025-08,West Midlands Police,West Midlands Police,-1.847899,52.591564,On or near Royal Park Close,E01009417,Birmingham 001A,Shoplifting,Awaiting court outcome,NaN
3,ca760e7e4a2d777c656807242187e8bf1b33e3457d222c...,2025-08,West Midlands Police,West Midlands Police,-1.847899,52.591564,On or near Royal Park Close,E01009417,Birmingham 001A,Shoplifting,Investigation complete; no suspect identified,NaN
4,0729cdd7e2bb71ce3388a94f1d18aff6ddcb244e4d9298...,2025-08,West Midlands Police,West Midlands Police,-1.849177,52.593202,On or near Russell Bank Road,E01009417,Birmingham 001A,Vehicle crime,Investigation complete; no suspect identified,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...
668103,34f3b6836640fb4f3090ac34324c1199fed419abba322a...,2024-08,West Midlands Police,West Midlands Police,-2.117890,52.570935,On or near Rowan Tree Drive,E01034315,Wolverhampton 035J,Violence and sexual offences,Investigation complete; no suspect identified,NaN
668104,831298360969695e1239756ff3d00d5e05b40859c1d618...,2024-08,West Midlands Police,West Midlands Police,-2.118019,52.569505,On or near Thompson Avenue,E01034315,Wolverhampton 035J,Violence and sexual offences,Investigation complete; no suspect identified,NaN
668105,c1f876d94affce28375d82528ab5f56e8487398c5aa725...,2024-08,West Midlands Police,West Midlands Police,-2.117890,52.570935,On or near Rowan Tree Drive,E01034315,Wolverhampton 035J,Violence and sexual offences,Local resolution,NaN
668106,1dbd69760b91c2afbea99fa1cc6d9b2c88911620166eb6...,2024-08,West Midlands Police,West Midlands Police,-2.121527,52.573817,On or near Waterside Close,E01034315,Wolverhampton 035J,Violence and sexual offences,Unable to prosecute suspect,NaN


**Reasoning**:
To extract 'Year' and 'Month' from the existing 'Month' column, I'll first convert the 'Month' column to a datetime format using `pd.to_datetime`. Then, I'll use the `.dt.year` and `.dt.month` accessors to create new 'Year' and 'Month' columns, respectively. Finally, I'll display the head of the DataFrame to verify the new columns.

In [9]:
combined_df['Year'] = pd.to_datetime(combined_df['Month']).dt.year
combined_df['Month_num'] = pd.to_datetime(combined_df['Month']).dt.month

combined_df.head(5)

,Crime ID,Month,Reported by,Falls within,Longitude,Latitude,Location,LSOA code,LSOA name,Crime type,Last outcome category,Context,Year,Month_num
0,1fc0f0d787b2f716930f49ff2eb9a49c55f327cf543aac...,2025-08,West Midlands Police,West Midlands Police,-1.847899,52.591564,On or near Royal Park Close,E01009417,Birmingham 001A,Criminal damage and arson,Unable to prosecute suspect,NaN,2025,8
1,626f61477f4c1c74fb96475907221d457ba73b887611bb...,2025-08,West Midlands Police,West Midlands Police,-1.847899,52.591564,On or near Royal Park Close,E01009417,Birmingham 001A,Drugs,Local resolution,NaN,2025,8
2,1122d890c2bc0e57a44a8b6ba5518ea9e98ee25eaf76b7...,2025-08,West Midlands Police,West Midlands Police,-1.847899,52.591564,On or near Royal Park Close,E01009417,Birmingham 001A,Shoplifting,Awaiting court outcome,NaN,2025,8
3,ca760e7e4a2d777c656807242187e8bf1b33e3457d222c...,2025-08,West Midlands Police,West Midlands Police,-1.847899,52.591564,On or near Royal Park Close,E01009417,Birmingham 001A,Shoplifting,Investigation complete; no suspect identified,NaN,2025,8
4,0729cdd7e2bb71ce3388a94f1d18aff6ddcb244e4d9298...,2025-08,West Midlands Police,West Midlands Police,-1.849177,52.593202,On or near Russell Bank Road,E01009417,Birmingham 001A,Vehicle crime,Investigation complete; no suspect identified,NaN,2025,8


In [10]:
# ==========================================
# STAGE 1: PLANNING - INTERNAL DATA VALIDATION
# ==========================================

print("--- Data Validation Report ---")

# 1. Row Count Validation
# We iterate through the original files just to count the rows and compare it to our master dataframe.
total_source_rows = 0
for file_path in all_files:
    # Reading just one column to speed up the row counting process
    df_temp = pd.read_csv(file_path, usecols=[0])
    total_source_rows += len(df_temp)

combined_rows = len(combined_df)

print(f"Total rows in source files: {total_source_rows}")
print(f"Total rows in combined DataFrame: {combined_rows}")

if total_source_rows == combined_rows:
    print("✅ PASS: Row counts perfectly match.")
else:
    print("❌ FAIL: Row counts do not match. Check your merging logic.")

print("\n------------------------------\n")

# 2. Missing Months Validation
# Extract the unique months and sort them
unique_months = combined_df['Month'].dropna().unique()
unique_months.sort()

month_count = len(unique_months)
expected_months = 24

print(f"Number of unique months found: {month_count}")
print(f"Expected number of months: {expected_months}")

if month_count == expected_months:
    print("✅ PASS: All 24 months are present.")
else:
    print("❌ FAIL: Missing or extra months detected.")

# 3. Date Range Check
first_month = unique_months[0]
last_month = unique_months[-1]

print(f"\nDate Range Found: {first_month} to {last_month}")
if first_month == '2023-12' and last_month == '2025-11':
    print("✅ PASS: Date range matches the brief (Dec 2023 - Nov 2025).")
else:
    print("⚠️ WARNING: Date range does not match the expected Dec 2023 to Nov 2025 range.")

--- Data Validation Report ---
Total rows in source files: 668108
Total rows in combined DataFrame: 668108
✅ PASS: Row counts perfectly match.

------------------------------

Number of unique months found: 24
Expected number of months: 24
✅ PASS: All 24 months are present.

Date Range Found: 2023-12 to 2025-11
✅ PASS: Date range matches the brief (Dec 2023 - Nov 2025).


In [11]:
# ==========================================
# STAGE 2: DATA PREPARATION - CLEANING NaNs
# ==========================================

print("--- Initial Missing Values Count ---")
print(combined_df.isnull().sum())

# 1. Drop the 'Context' column
# Justification: This column is almost always completely empty in the data.police.uk dataset.
if 'Context' in combined_df.columns:
    combined_df_cleaned = combined_df.drop(columns=['Context'])

# 2. Handle missing 'Crime ID' and 'Last outcome category'
# Justification: ASB often lacks a Crime ID and formal outcome. We fill these with descriptive placeholders
# rather than dropping the rows, so we don't lose valuable ASB trend data.
combined_df_cleaned['Crime ID'] = combined_df_cleaned['Crime ID'].fillna('No ID Provided')
combined_df_cleaned['Last outcome category'] = combined_df_cleaned['Last outcome category'].fillna('No Outcome Recorded')

# 3. Handle missing Location/LSOA data
# Justification: Sometimes police withhold exact locations for privacy (anonymisation).
# We flag these rather than deleting the crime record.
combined_df_cleaned['LSOA name'] = combined_df_cleaned['LSOA name'].fillna('Location Anonymised/Unknown')
combined_df_cleaned['LSOA code'] = combined_df_cleaned['LSOA code'].fillna('Unknown')

# Note on Longitude/Latitude: We will leave these as NaN for now. If you do geospatial mapping later,
# you can filter out rows where coordinates are missing (e.g., using .dropna(subset=['Longitude', 'Latitude'])).

print("\n--- Missing Values After Cleaning ---")
print(combined_df_cleaned.isnull().sum())

# Let's peek at the cleaned data
combined_df_cleaned.head()

--- Initial Missing Values Count ---
Crime ID                  51667
Month                         0
Reported by                   0
Falls within                  0
Longitude                     0
Latitude                      0
Location                      0
LSOA code                     0
LSOA name                     0
Crime type                    0
Last outcome category     51667
Context                  668108
Year                          0
Month_num                     0
dtype: int64

--- Missing Values After Cleaning ---
Crime ID                 0
Month                    0
Reported by              0
Falls within             0
Longitude                0
Latitude                 0
Location                 0
LSOA code                0
LSOA name                0
Crime type               0
Last outcome category    0
Year                     0
Month_num                0
dtype: int64


,Crime ID,Month,Reported by,Falls within,Longitude,Latitude,Location,LSOA code,LSOA name,Crime type,Last outcome category,Year,Month_num
0,1fc0f0d787b2f716930f49ff2eb9a49c55f327cf543aac...,2025-08,West Midlands Police,West Midlands Police,-1.847899,52.591564,On or near Royal Park Close,E01009417,Birmingham 001A,Criminal damage and arson,Unable to prosecute suspect,2025,8
1,626f61477f4c1c74fb96475907221d457ba73b887611bb...,2025-08,West Midlands Police,West Midlands Police,-1.847899,52.591564,On or near Royal Park Close,E01009417,Birmingham 001A,Drugs,Local resolution,2025,8
2,1122d890c2bc0e57a44a8b6ba5518ea9e98ee25eaf76b7...,2025-08,West Midlands Police,West Midlands Police,-1.847899,52.591564,On or near Royal Park Close,E01009417,Birmingham 001A,Shoplifting,Awaiting court outcome,2025,8
3,ca760e7e4a2d777c656807242187e8bf1b33e3457d222c...,2025-08,West Midlands Police,West Midlands Police,-1.847899,52.591564,On or near Royal Park Close,E01009417,Birmingham 001A,Shoplifting,Investigation complete; no suspect identified,2025,8
4,0729cdd7e2bb71ce3388a94f1d18aff6ddcb244e4d9298...,2025-08,West Midlands Police,West Midlands Police,-1.849177,52.593202,On or near Russell Bank Road,E01009417,Birmingham 001A,Vehicle crime,Investigation complete; no suspect identified,2025,8


In [12]:
unique_crime_ids = combined_df_cleaned['Crime ID'].nunique()
total_rows = len(combined_df_cleaned)

print(f"Total rows in DataFrame: {total_rows}")
print(f"Number of unique Crime IDs: {unique_crime_ids}")

if unique_crime_ids == total_rows:
    print("✅ All rows have unique Crime IDs.")
else:
    print("❌ There are duplicate Crime IDs.")
    duplicate_crime_ids = combined_df_cleaned[combined_df_cleaned.duplicated(subset=['Crime ID'], keep=False)]
    num_duplicates = duplicate_crime_ids['Crime ID'].nunique()
    print(f"Number of unique Crime IDs with duplicates: {num_duplicates}")
    print(f"Total number of rows involved in duplicates: {len(duplicate_crime_ids)}")
    print("Here are the first 5 rows of Crime IDs that are duplicates:")
    display(duplicate_crime_ids.head())

Total rows in DataFrame: 668108
Number of unique Crime IDs: 616442
❌ There are duplicate Crime IDs.
Number of unique Crime IDs with duplicates: 1
Total number of rows involved in duplicates: 51667
Here are the first 5 rows of Crime IDs that are duplicates:


,Crime ID,Month,Reported by,Falls within,Longitude,Latitude,Location,LSOA code,LSOA name,Crime type,Last outcome category,Year,Month_num
25,No ID Provided,2025-08,West Midlands Police,West Midlands Police,-1.831228,52.592809,On or near Crockford Drive,E01009415,Birmingham 002A,Anti-social behaviour,No Outcome Recorded,2025,8
88,No ID Provided,2025-08,West Midlands Police,West Midlands Police,-1.823466,52.581219,On or near Trinity Road,E01009431,Birmingham 003E,Anti-social behaviour,No Outcome Recorded,2025,8
89,No ID Provided,2025-08,West Midlands Police,West Midlands Police,-1.827047,52.582501,On or near Jordan Close,E01009431,Birmingham 003E,Anti-social behaviour,No Outcome Recorded,2025,8
119,No ID Provided,2025-08,West Midlands Police,West Midlands Police,-1.834052,52.556566,On or near Murray Court,E01009468,Birmingham 004F,Anti-social behaviour,No Outcome Recorded,2025,8
120,No ID Provided,2025-08,West Midlands Police,West Midlands Police,-1.843323,52.554223,On or near Stonehouse Road,E01009468,Birmingham 004F,Anti-social behaviour,No Outcome Recorded,2025,8


# Task
Review data types of all columns in `combined_df_cleaned` and convert the 'Month' column to datetime objects. Additionally, identify suitable string columns and convert them to categorical data types to optimize memory usage and improve performance.

## Review and Convert Data Types

### Subtask:
Inspect the data types of all columns in `combined_df_cleaned` and convert them to more appropriate types where necessary. Ensure the 'Month' column is a datetime object and convert suitable string columns to categorical types to optimize memory and performance.


**Reasoning**:
First, I'll display the current data types of `combined_df_cleaned` to understand their initial state before any conversions.



In [14]:
print("Current Data Types:")
print(combined_df_cleaned.dtypes)

Current Data Types:
Crime ID                  object
Month                     object
Reported by               object
Falls within              object
Longitude                float64
Latitude                 float64
Location                  object
LSOA code                 object
LSOA name                 object
Crime type                object
Last outcome category     object
Year                       int32
Month_num                  int32
dtype: object


**Reasoning**:
Now that the current data types have been inspected, I will proceed to convert the 'Month' column to datetime and suitable string columns to categorical types to optimize the DataFrame.



In [15]:
combined_df_cleaned['Month'] = pd.to_datetime(combined_df_cleaned['Month'])

# Identify columns that are suitable for conversion to 'category' type
categorical_cols = [
    'Reported by',
    'Falls within',
    'Location',
    'LSOA code',
    'LSOA name',
    'Crime type',
    'Last outcome category'
]

for col in categorical_cols:
    if col in combined_df_cleaned.columns and combined_df_cleaned[col].dtype == 'object':
        combined_df_cleaned[col] = combined_df_cleaned[col].astype('category')

print("\nData Types After Conversion:")
print(combined_df_cleaned.dtypes)


Data Types After Conversion:
Crime ID                         object
Month                    datetime64[ns]
Reported by                    category
Falls within                   category
Longitude                       float64
Latitude                        float64
Location                       category
LSOA code                      category
LSOA name                      category
Crime type                     category
Last outcome category          category
Year                              int32
Month_num                         int32
dtype: object


## Categorical Data Consistency Check

### Subtask:
Examine unique values in key categorical columns like 'Crime type', 'Reported by', 'Falls within', 'Last outcome category', and 'LSOA name' to identify any inconsistencies, typos, or variations that might require standardisation or grouping.


**Reasoning**:
To identify inconsistencies and variations in the categorical columns, I will iterate through the specified columns and display their unique values and value counts. For 'LSOA name', which can have many unique values, I will limit the display of value counts to the top 50 for readability.



In [16]:
key_categorical_cols = [
    'Crime type',
    'Reported by',
    'Falls within',
    'Last outcome category',
    'LSOA name'
]

print("--- Categorical Data Consistency Check ---")

for col in key_categorical_cols:
    print(f"\nColumn: {col}")
    print(f"Unique values ({combined_df_cleaned[col].nunique()}): {combined_df_cleaned[col].unique()}")

    if col == 'LSOA name':
        print("Top 50 Value Counts:")
        print(combined_df_cleaned[col].value_counts().head(50))
    else:
        print("Value Counts:")
        print(combined_df_cleaned[col].value_counts())


--- Categorical Data Consistency Check ---

Column: Crime type
Unique values (14): ['Criminal damage and arson', 'Drugs', 'Shoplifting', 'Vehicle crime', 'Violence and sexual offences', ..., 'Anti-social behaviour', 'Bicycle theft', 'Robbery', 'Possession of weapons', 'Theft from the person']
Length: 14
Categories (14, object): ['Anti-social behaviour', 'Bicycle theft', 'Burglary', 'Criminal damage and arson',
                          ..., 'Shoplifting', 'Theft from the person', 'Vehicle crime',
                          'Violence and sexual offences']
Value Counts:
Crime type
Violence and sexual offences    266582
Shoplifting                      62492
Vehicle crime                    59127
Anti-social behaviour            51667
Criminal damage and arson        50311
Other theft                      41491
Public order                     37459
Burglary                         32222
Drugs                            17005
Robbery                          14162
Other crime              